### PR TITLE
Capacity tear sheet

### PR DESCRIPTION
**Pull request by a-campbell**
There are two main analyses here: constraining ticker and a capacity sweep.

The constraining ticker analysis asks "What is the maximum capital base I could have deployed this strategy on without ever consuming more than X% of the {average, 10th_percentile, 90th_percentile} daily volume for each traded name.

The capacity sweep asks "How does my strategy's Sharpe ratio decay as I increase capital base and incur greater slippage on my trades?"

Another change of note is the inclusion of market data in this tear sheet. Since we will likely want to use market data for future tear sheets, we should be sure that we like the Panel data format we are asking for in this API change.

screenshot 2016-04-01 17 19 56
screenshot 2016-04-01 17 22 48
screenshot 2016-04-01 17 22 58

Wondering if it is useful to display the most liquid names in the constraining ticker analysis?

### CODE REVIEW

**Comment by a-campbell** [comment]
I'm not sure how to make Travis accept `get_ipython()` as an unbound function in utils.py

**Comment by twiecki** [comment]
```
pyfolio/plotting.py:1269:21: F821 undefined name 'xrange'
pyfolio/tears.py:683:65: W291 trailing whitespace
pyfolio/utils.py:484:9: F821 undefined name 'get_ipython'
```

The first is easy to fix either with `six` or just using `range`.
Third can be fixed with by ending the line with  `# noqa` to disable flake.

File: pyfolio/capacity.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+from . import pos
+from . import timeseries
+
+DISPLAY_UNIT = 1000000
+
+
+def get_assets_dollar_volume(positions, market_data,
+                             only_held_days=True, last_n_days=None):
+    """
+    Computes descriptive statsitics for the daily dollar
```
Comment (twiecki) [inline pyfolio/capacity.py]: `statsitics`

File: pyfolio/capacity.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        Compute max position allocation and dollar volume for only
+        the last N days of the backtest
+
+    Returns
+    -------
+    vol_analysis : pd.DataFrame
+        Index of traded tickers. Columns for 10th percentile, 90th percentile,
+        and mean daily traded volume, as well as max position concentration.
+    """
+    DV = market_data['volume'] * market_data['price']
```
Comment (twiecki) [inline pyfolio/capacity.py]: `dollar_value`, 'vwap'?

File: pyfolio/capacity.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    positions_alloc = pos.get_percent_alloc(positions)
+    positions_alloc = positions_alloc.drop('cash', axis=1)
+
+    if last_n_days is not None:
+        positions_alloc = positions_alloc.iloc[-last_n_days:]
+        DV = DV.iloc[-last_n_days:]
+
+    if only_held_days:
+        for name, ts in positions_alloc.iteritems():
```
Comment (twiecki) [inline pyfolio/capacity.py]: `ts`? `pos_alloc`?

File: pyfolio/capacity.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            DV[name] = DV.loc[ts[ts != 0].index, name]
+
+    max_exposure_per_ticker = abs(positions_alloc).max()
```
Comment (twiecki) [inline pyfolio/capacity.py]: Can be `positions_alloc.abs().max()`

File: pyfolio/capacity.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
+from __future__ import division
+import pandas as pd
+import numpy as np
+from . import pos
+from . import timeseries
+
+DISPLAY_UNIT = 1000000
```
Comment (twiecki) [inline pyfolio/capacity.py]: Why this value?

File: pyfolio/capacity.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+def get_portfolio_size_constraints(vol_analysis, daily_vol_limit=0.2):
+    """
+    Finds max portfolio size at daily volume limit given at different
+    slices of daily dollar volume distributions.
+
+    Parameters
+    ----------
+    vol_analysis : pd.DataFrame
+        See output of get_assets_dollar_volume
+    daily_vol_limit : float
```
Comment (twiecki) [inline pyfolio/capacity.py]: `(optional)`

File: pyfolio/capacity.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        consume.
+
+    Returns
+    -------
+    constraint_tickers : pd.DataFrame
+        Algo capacity and constraining ticker at various parts of
+        daily volume distribution.
+    """
+    constraints = pd.DataFrame()
+    constraints['max_capacity_at_ADTV'] = np.round(
```
Comment (twiecki) [inline pyfolio/capacity.py]: Any reason to use numpy round instead of `DataFrame.round`?

File: pyfolio/plotting.py [hunk 8]
```diff
@@ -1260,6 +1261,33 @@ def plot_slippage_sensitivity(returns, transactions, positions,
     return ax
 
 
+def plot_capacity_sweep(returns, daily_txn, bt_starting_capital,
+                        min_pv=100000, max_pv=300000000, step_size=1000000,
```
Comment (twiecki) [inline pyfolio/plotting.py]: Might be more readable with `1E5` etc.

File: pyfolio/plotting.py [hunk 9]
```diff
@@ -1260,6 +1261,33 @@ def plot_slippage_sensitivity(returns, transactions, positions,
...
+                        ax=None):
+
+    captial_base_sweep = pd.Series()
+    for start_pv in xrange(min_pv, max_pv, step_size):
```
Comment (twiecki) [inline pyfolio/plotting.py]: `xrange` is depracated in python 3. Either use `six.range` or just `range` and eat the memory cost in python 2 which shouldn't be a big deal here.

File: pyfolio/tears.py [hunk 10]
```diff
@@ -102,6 +104,10 @@ def create_full_tear_sheet(returns,
             2004-01-09 12:18:01    483      324.12   'AAPL'
             2004-01-09 12:18:01    122      83.10    'MSFT'
             2004-01-13 14:12:23    -75      340.43   'AAPL'
+    market_data : pd.Panel, optional
```
Comment (twiecki) [inline pyfolio/tears.py]: `market_ohlcv`? A bit cryptic so I'm not sure if that's really better.

File: pyfolio/utils.py [hunk 11]
```diff
@@ -470,3 +470,22 @@ def get_symbol_rets(symbol, start=None, end=None):
     return SETTINGS['returns_func'](symbol,
                                     start=start,
                                     end=end)
+
+
+def print_table(table, name=None, fmt=None):
```
Comment (twiecki) [inline pyfolio/utils.py]: doc string would be helpful.

File: pyfolio/tears.py [hunk 12]
```diff
@@ -669,6 +680,86 @@ def create_interesting_times_tear_sheet(
 
 
 @plotting_context
+def create_capacity_tear_sheet(returns, positions, transactions, 
+                               market_data,
+                               daily_vol_limit=0.2, last_n_days=125):
```
Comment (twiecki) [inline pyfolio/tears.py]: `APPROX_BDAYS_PER_MONTH * 6`

File: pyfolio/tears.py [hunk 13]
```diff
@@ -669,6 +680,86 @@ def create_interesting_times_tear_sheet(
...
+
+    whole_bt = capacity.get_assets_dollar_volume(positions,
+                                                 market_data,
+                                                 only_held_days=True)
+    n_days = capacity.get_assets_dollar_volume(positions,
+                                               market_data,
+                                               only_held_days=True,
+                                               last_n_days=last_n_days)
+    for title, data in [('Whole backtest', whole_bt),
+                        ('Last six months:', n_days)]:
```
Comment (twiecki) [inline pyfolio/tears.py]: If it's a kwarg won't always be six months.

File: pyfolio/tears.py [hunk 14]
```diff
@@ -669,6 +680,86 @@ def create_interesting_times_tear_sheet(
...
+        positions DataFrames
+    daily_vol_limit : float
+        Max proportion of any daily bar that the startegy is allowed to
+        consume in the constraining ticker analysis.
+    last_n_days : integer
+        Compute max position allocation and dollar volume for only
+        the last N days of the backtest
+    """
+
+    whole_bt = capacity.get_assets_dollar_volume(positions,
```
Comment (twiecki) [inline pyfolio/tears.py]: Could be more readable if it'd be called `asset_dollar_volume_whole_bt`.

File: pyfolio/tears.py [hunk 15]
```diff
@@ -669,6 +680,86 @@ def create_interesting_times_tear_sheet(
...
+                                                 market_data,
+                                                 only_held_days=True)
+    n_days = capacity.get_assets_dollar_volume(positions,
+                                               market_data,
+                                               only_held_days=True,
+                                               last_n_days=last_n_days)
+    for title, data in [('Whole backtest', whole_bt),
```
Comment (twiecki) [inline pyfolio/tears.py]: In general good to find more descriptive names than `data`.

File: pyfolio/tears.py [hunk 16]
```diff
@@ -669,6 +680,86 @@ def create_interesting_times_tear_sheet(
...
+            data.sort('avg_daily_dv_$mm', ascending=True).head(8))
+
+        constraints = capacity.get_portfolio_size_constraints(
+            data, daily_vol_limit=daily_vol_limit)
+        constraining_tickers = capacity.get_constraining_tickers(constraints)
+        print("Max capacity at %s%% pct daily bar consumption limit" %
+              (daily_vol_limit * 100))
+        utils.print_table(constraining_tickers)
+
+    illiquid = whole_bt[whole_bt['avg_daily_dv_$mm'] < 1]
```
Comment (twiecki) [inline pyfolio/tears.py]: Maybe also use a const here instead of `1` to make this more clear, like `MIN_DAILY_DV_MM`.

File: pyfolio/tears.py [hunk 17]
```diff
@@ -669,6 +680,86 @@ def create_interesting_times_tear_sheet(
...
+    if len(illiquid) > 0:
+        fig = plt.figure(figsize=(14, 2 * 6))
+        gs = gridspec.GridSpec(2, 1, wspace=0.5, hspace=0.5)
+        ax_illiquid = plt.subplot(gs[0, :])
+        ax_capacity_sweep = plt.subplot(gs[1, :])
+
+        pos_alloc = pos.get_percent_alloc(positions)
+        (pos_alloc[illiquid.index.values] * 100).plot(ax=ax_illiquid)
```
Comment (twiecki) [inline pyfolio/tears.py]: is `.values` required?

File: pyfolio/capacity.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
+from __future__ import division
+import pandas as pd
+import numpy as np
+from . import pos
+from . import timeseries
+
+DISPLAY_UNIT = 1000000
+
+
+def get_assets_dollar_volume(positions, market_data,
```
Comment (twiecki) [inline pyfolio/capacity.py]: Would it make things more readable if we ommited `dollar` and only called it volume (we can't be sure it's dollars to begin with). Then, `DV` could become `vol` which is also more readable.

**Comment by twiecki** [comment]
I haven't really followed the development of the capacity analysis so coming to this with fresh eyes. It's possible that this is the best to display the information we want to extract but overall I find it a bit hard to extract actionable insights. The below points are thus mostly naive questions rather than fully baked feedback on what would be better. 
 
* I find the tables a bit hard to read but not quite sure how to improve that, especially if many values are `0.00`. Perhaps a box plot?
* The alloc plot of the illiquid tickers also feels a bit uninformative unless I'm missing what should jump out here, and the 1mm$ threshold seems a bit arbitrary. I'd also drop the most liquid tickers analysis as you suggest.
* The capacity sweep plot is really nice. However, putting things together, the max capacity at 20% volume at the 90th percentile (which seems pretty aggressive) suggests a max capital base of $50,000. However, the capacity sweep suggests I can trade this strategy at $20MM easily.
* Have you considered a plot that shows for each ticker (or the N least liquid ones), over time, how much % of the vol was I taking up? That would tell me if there perhaps was just one squeeze or whether I was constantly using up most of the stocks liquidity.

File: pyfolio/capacity.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+            DV[name] = DV.loc[ts[ts != 0].index, name]
+
+    max_exposure_per_ticker = abs(positions_alloc).max()
+
+    vol_analysis = pd.DataFrame()
+    vol_analysis['algo_max_exposure'] = max_exposure_per_ticker
+    vol_analysis.loc[:, 'avg_daily_dv_$mm'] = np.round(
+        DV.mean() / DISPLAY_UNIT, 2)
+    vol_analysis['10th_%_daily_dv_$mm'] = np.round(DV.apply(
+        lambda x: np.nanpercentile(x, 10)) / DISPLAY_UNIT, 2)
```
Comment (twiecki) [inline pyfolio/capacity.py]: `DV.quantile(q=0.1)`

File: pyfolio/capacity.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+    vol_analysis['90th_%_daily_dv_$mm'] = np.round(DV.apply(
+        lambda x: np.nanpercentile(x, 90)) / DISPLAY_UNIT, 2)
```
Comment (twiecki) [inline pyfolio/capacity.py]: This whole block would look much better if the round and display unit conversion would be done at the end over the columns you want.

File: pyfolio/capacity.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+    Sums the absolute value of shares traded in each name on each day.
+    Adds columns containing the closing price and total daily volume for
+    each day-ticker combination.
+
+    Parameters
+    ----------
+    transactions : pd.DataFrame
+        Prices and amounts of executed trades. One row per trade.
+        - See full explanation in tears.create_full_tear_sheet
+    market_data : pd.DataFrame
```
Comment (twiecki) [inline pyfolio/capacity.py]: `Panel`?

File: pyfolio/capacity.py [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+    -------
+    txn_daily : pd.DataFrame
+        Daily totals for transacted shares in each traded name.
+        price and volume columns for close price and daily volume for
+        the corresponding ticker, respectively.
+    """
+    transactions.index.name = 'date'
+    txn_daily = pd.DataFrame(transactions.assign(
+        amount=abs(transactions.amount)).groupby(
+        ['symbol', pd.TimeGrouper('D')]).sum()['amount'])
```
Comment (twiecki) [inline pyfolio/capacity.py]: Neat!

File: pyfolio/capacity.py [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+
+    txn_daily = txn_daily.reset_index().set_index('date')
+
+    return txn_daily
+
+
+def apply_slippage_penalty(returns, txn_daily, simulate_starting_capital,
+                           backtest_starting_capital, impact=0.1):
+    """
+    Applies quadratic volumeshare slippage model to daily return based
```
Comment (twiecki) [inline pyfolio/capacity.py]: returns

File: pyfolio/plotting.py [hunk 24]
```diff
@@ -1260,6 +1261,33 @@ def plot_slippage_sensitivity(returns, transactions, positions,
...
+        adj_ret = capacity.apply_slippage_penalty(returns,
+                                                  daily_txn,
+                                                  start_pv,
+                                                  bt_starting_capital)
+        sharpe = timeseries.sharpe_ratio(adj_ret)
+        if sharpe < -1:
+            break
+        captial_base_sweep.loc[start_pv] = sharpe
+    captial_base_sweep.index = captial_base_sweep.index / 1000000.
```
Comment (twiecki) [inline pyfolio/plotting.py]: Use the same constant as above?

File: pyfolio/capacity.py [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+    mult = simulate_starting_capital / backtest_starting_capital
+    simulate_traded_shares = abs(mult * txn_daily.amount)
+    simulate_traded_dollars = txn_daily.price * simulate_traded_shares
+
+    penalties = (simulate_traded_shares / txn_daily.volume)**2 \
+        * impact * simulate_traded_dollars
+
+    daily_penalty = penalties.resample('D', how='sum')
+    daily_penalty = daily_penalty.reindex(returns.index).fillna(0)
+
```
Comment (twiecki) [inline pyfolio/capacity.py]: Very readable :+1:

File: pyfolio/capacity.py [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+    Returns
+    -------
+    adj_returns : pd.Series
+        Slippage penalty adjusted daily returns.
+    """
+    mult = simulate_starting_capital / backtest_starting_capital
+    simulate_traded_shares = abs(mult * txn_daily.amount)
+    simulate_traded_dollars = txn_daily.price * simulate_traded_shares
+
+    penalties = (simulate_traded_shares / txn_daily.volume)**2 \
```
Comment (twiecki) [inline pyfolio/capacity.py]: Would be even clearer with a variable in-between: `simulated_pct_volume_used = simulate_traded_shares / txn_daily.volume`.

File: pyfolio/capacity.py [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+        * impact * simulate_traded_dollars
+
+    daily_penalty = penalties.resample('D', how='sum')
+    daily_penalty = daily_penalty.reindex(returns.index).fillna(0)
+
+    # Since we are scaling the numerator of the penalties linearly
+    # by capital base, it makes the most sense to scale the denominator
+    # similarly. In other words, since we aren't applying compounding to
+    # simulate_traded_shares, we shouldn't apply compounding to pv.
+    pv = timeseries.cum_returns(
```
Comment (twiecki) [inline pyfolio/capacity.py]: `pv`?

File: pyfolio/capacity.py [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+        returns, starting_value=backtest_starting_capital) * mult
+
+    adj_returns = returns - (daily_penalty / pv)
```
Comment (twiecki) [inline pyfolio/capacity.py]: `daily_penalty` is in dollars, no?

File: pyfolio/plotting.py [hunk 29]
```diff
@@ -909,24 +910,24 @@ def show_and_plot_top_positions(returns, positions_alloc,
     if show_and_plot == 1 or show_and_plot == 2:
         print("\n")
         print('Top 10 long positions of all time (and max%)')
-        print(pd.DataFrame(df_top_long).index.values)
+        print([str(sym) for sym in df_top_long.index.values])
```
Comment (twiecki) [inline pyfolio/plotting.py]: Probably don't need the `.values`.

File: pyfolio/tears.py [hunk 30]
```diff
@@ -102,6 +104,10 @@ def create_full_tear_sheet(returns,
             2004-01-09 12:18:01    483      324.12   'AAPL'
             2004-01-09 12:18:01    122      83.10    'MSFT'
             2004-01-13 14:12:23    -75      340.43   'AAPL'
+    market_data : pd.Panel, optional
+        Panel containing open, high, low, close, and volume DataFrames.
```
Comment (twiecki) [inline pyfolio/tears.py]: If it's a panel I'd specify which are major, minor and item axes.

**Comment by twiecki** [comment]
OK, done with this round.

File: pyfolio/tears.py (hunk 10 above)
Comment (a-campbell) [inline pyfolio/tears.py]: All we really need is a items axis including 'price' and 'volume'.

**Comment by a-campbell** [comment]
@twiecki I think this should be good for another look. I updated the screenshots in the original PR post.

**Comment by twiecki** [comment]
I think this looks great. You'll need to rebase as we have `print_table` in master now.

**Comment by a-campbell** [comment]
All squashed and rebased now.

**Comment by twiecki** [comment]
:+1:

**Pull request merged**